ncclient/operations/retrieve.py

```diff
@@ -165,6 +171,141 @@ def request(self, source, filter=None, with_defaults=None):
             )
         return self._request(node)
 
+
+class GetData(RPC):
+    """RFC8526 implementation of <get-data> RPC"""
+
+    REPLY_CLS = GetReply
+    """See :class:`GetReply`."""
+
+    def request(self, datastore, filter=None, config_filter=False,
+            origin_filters=[], negated_origin_filters=[], max_depth=None,
+            with_origin=None, with_defaults=None, strip_ns=False):
+        """Construct the request for the get-data RPC
+
+        See RFC8526 for detailed explanations of all posssible input
+        parameters as is a 1:1 relationship in most cases.
+
+        Outside of the specification, if the caller specifies
+        strip_ns=True, then the 'datastore' element will not contain a
+        prefix xmlns declaration in addition the text content will not
+        contain a prepended prefix.  This is necessary for some
+        implementations that do not handle the namespace/prefix
+        declaration properly.
+
+        Args:
+            datastore: A valid _types.Datastore enum value representing
+                a datastore to query [mandatory]
+            filter: A tuple, list or string representing filter data
+            config_filter: A boolean value representing to only return
+                config=true (r/w) nodes from the requested datastore
+            origin_filters: A list of _types.Origin values
+            negated_origin_filters: A list of _types.Origin values
+            max_depth: None for 'unbounded', otherwise an integer value
+                between 1-65535
+            with_origin: A boolean value representing to return origin
+                metadata in responses
+            with_defaults: A _types.WithDefaults enum value
+                representing to return default data according to
+                RFC6243
+            strip_ns: A boolean value to indicate skipping namespace
+                qualification of the datastore node
+        Returns:
+            An ncclient.xml_.NCElement representing the serialized
+            NETCONF RPC request.
+        """
+
+        if not isinstance(datastore, Datastore):
```

Comment:
Did you consider keeping parity with the existing get_config call, which identifies a target/datastore just using a string? Or enabling both? (As per your changes to filtering to maintain b/c.)

Reply:
(Asking as if developers use both get-data & get-config, parameters will differ, but also aware that if a device is NMDA-compliant a developer my not use get-config as well as get-data.)

Reply:
My original thoughts here were:

* To keep backwards compatibility on existing code paths
* To introduce a more robust API design that other operations would adhere to in subsequent commits by introducing in a backwards compatible fashion
* Eventually look to deprecate the use of opaque strings for such uses

While the passing and direct marshaling of string values is expansible without code changes should an implementation support alternate non-standard datastores, the reason I didn't introduce accepting strings here was to have some API enforcement vs. blind marshaling.

However this raises additional work that needs accounting for - for the expansibility of the base class `Datastore`.  We want the various device handlers to be able to extend and operate on this class much like one would build off the base ds:datastore identity via YANG models.

Let me loop back around to this to account for the expansibility of enum expansions and worst case, can open this up to accept opaque strings as it is today if we feel we want to maintain this loose typing.

Reply:
@earies, what if you did a version with the loose typing, and then we consider either making more type-safe APIs in a subsequent version? The subsequent version would be more expansive and hence backwards-compatibility breaking.

We could either:

1. Declare a 1.0.0 with the loose typing changes and then move onto a 2.x.y; or
2. Have the loose typing changes as the last 0.x.y release, maybe 0.7.0, and then make a 1.0.0 with the more expansive changes

I prefer option #2, which would allow us to keep 0.7.x as a branch for the loosely typed APIs.

Thoughts?